File: src/objects/encoder.cpp
#include "encoder.hpp"

char StrEncoder::get_type_id(const Object& object) {
  if (object.type == Object::Type::text) {
    return 1;
  } 

  if (object.type == Object::Type::loginAttempt) {
    return 2;
  }

  if (object.type == Object::Type::returnCode) {
    return 3;
  }

  if (object.type == Object::Type::command) {
    return 4;
  }
  return 0;
}

Object::Type StrEncoder::from_type_id(char id) {
  if (id == 1) {
    return Object::Type::text;
  }

  if (id == 2) {
    return Object::Type::loginAttempt;
  }

  if (id == 3) {
    return Object::Type::returnCode;
  }

  if (id == 4) {
    return Object::Type::command;
  }

  return Object::Type::unknown;
}

template<typename T>
Encoder::bytes StrEncoder::to_bytes(T id) {
  const auto size = sizeof(T);
  char result[size];
    
  std::copy(
    reinterpret_cast<char*>(&id),
    reinterpret_cast<char*>(&id) + size,
    &result[0]
  );
  bytes b(result);
  b.resize(size);
  return b;
}

template<typename T>
T StrEncoder::from_bytes(const Encoder::bytes& b) {
  const auto size = sizeof(T);
  T result;
  const char* r_bytes = b.c_str();

  std::copy(
    &r_bytes[0],
    &r_bytes[0] + size,
    reinterpret_cast<char*>(&result)
  );

  return result;
}

#include <algorithm> // reverse
std::string remove_zeroes(std::string s) {
  std::string result;
  std::reverse(s.begin(), s.end());
  auto ptr = reinterpret_cast<const unsigned char*>(&s[0]);

  int carry = 0;
  for (int i = 0; i < s.size(); ++i) {
    int x = ptr[i];
    carry = carry * 256 + x;
    while (carry >= 255) {
      result.push_back(carry % 255 + 1);
      carry /= 255;
    }
  }

  if (carry != 0) {
    result.push_back(carry + 1);
  }

  return result;
}

std::string add_zeroes(std::string s) {
  std::string result;
  std::reverse(s.begin(), s.end());
  auto ptr = reinterpret_cast<const unsigned char*>(&s[0]);

  int carry = 0;

  for (int i = 0; i < s.size(); ++i) {
    int x = ptr[i];
    carry = carry * 255 + (x - 1);
    while (carry >= 256) {
      result.push_back(carry % 256);
      carry /= 256;
    }
  }

  if (carry != 0) {
    result.push_back(carry);
  }

  return result;
}


Encoder::bytes StrEncoder::encode(const Object& object) {
  std::string s = "";
  s.push_back(get_type_id(object));
  s.push_back(object.attributes);
  if (object.has_id())
    s += to_bytes(object.id);

  if (object.has_author())
    s += to_bytes(object.author);

  if (object.has_timestamp())
    s += to_bytes(object.timestamp);

  if (object.has_thread())
    s += to_bytes(object.thread);

  if (object.has_reply())
    s += to_bytes(object.reply);

  if (object.has_prev())
    s += to_bytes(object.prev);

  if (object.has_next())
    s += to_bytes(object.next);

  if (object.has_return_code())
    s += to_bytes(object.code);

  s += object.content;
  s.push_back(1);
  s = remove_zeroes(s);
  return s;
}

Object StrEncoder::decode(Encoder::bytes bytes) {
  Object object;
  bytes = add_zeroes(bytes);

  char type = bytes[0];
  object.type = from_type_id(type);
  object.attributes = bytes[1];

  int ptr = 2;
  if (object.has_id()) {
    object.id = from_bytes<int>(bytes.substr(ptr, sizeof(int)));
    ptr += sizeof(int);
  }

  if (object.has_author()) {
    object.author = from_bytes<int>(bytes.substr(ptr, sizeof(int)));
    ptr += sizeof(int);
  }
    
  if (object.has_timestamp()) {
    object.timestamp = from_bytes<int64_t>(bytes.substr(ptr, sizeof(int64_t)));
    ptr += sizeof(int64_t);
  }

  if (object.has_thread()) {
    object.thread = from_bytes<int>(bytes.substr(ptr, sizeof(int)));
    ptr += sizeof(int);
  }

  if (object.has_reply()) {
    object.reply = from_bytes<int>(bytes.substr(ptr, sizeof(int)));
    ptr += sizeof(int);
  }

  if (object.has_prev()) {
    object.prev = from_bytes<int>(bytes.substr(ptr, sizeof(int)));
    ptr += sizeof(int);
  }

  if (object.has_next()) {
    object.next = from_bytes<int>(bytes.substr(ptr, sizeof(int)));
    ptr += sizeof(int);
  }

  if (object.has_return_code()) {
    object.code = from_bytes<int>(bytes.substr(ptr, sizeof(int)));
    ptr += sizeof(int);
  }

  for (int i = ptr; i < bytes.size(); ++i) {
    object.content += bytes[i];
  }
  object.content.pop_back();
  
  return object;
}

File: python-gui/src/encoder.py
import numpy as np

def get_bit(var, order):
    return (var >> order) % 2

def set_bit(var, order, value):
    if (get_bit(var, order) != value):
        var ^= (1 << order)

    return var

class Object:
    content = ""
    attributes = dict()
    object_type = "unknown"

    def set_value(self, attribute, value):
        self.attributes[attribute] = value

    def get_value(self, attribute):
        if attribute not in self.attributes:
            return None

        return self.attributes[attribute]

class Encoder:
    def get_type_id(self, type_alias):
        type_id = {
            "text" : 1,
            "loginAttempt" : 2,
            "returnCode" : 3,
            "command" : 4,
        }

        if type_alias not in type_id:
            return 0
        
        return type_id[type_alias]

    def get_attribute_order(self, attribute_alias):
        attribute_order = {
            "id" : 0,
            "author" : 1,
            "timestamp" : 2,
            "thread" : 3,
            "reply" : 4,
            "prev" : 5, 
            "next" : 6,
            "returnCode" : 7
        }
        return attribute_order[attribute_alias]

    def get_encoded_attributes(self, obj):
        attributes_encoded = 0
        for key in obj.attributes:
            attributes_encoded = set_bit(
                attributes_encoded, self.get_attribute_order(key), 1
            )
        b = attributes_encoded.to_bytes(1, "big")

        if "id" in obj.attributes:
            b += obj.attributes["id"].to_bytes(4, "big")

        if "author" in obj.attributes:
            b += obj.attributes["author"].to_bytes(4, "big")

        if "timestamp" in obj.attributes:
            b += obj.attributes["timestamp"].to_bytes(8, "big")

        if "thread" in obj.attributes:
            b += obj.attributes["thread"].to_bytes(4, "big")

        if "reply" in obj.attributes:
            b += obj.attributes["reply"].to_bytes(4, "big")

        if "prev" in obj.attributes:
            b += obj.attributes["prev"].to_bytes(4, "big")
        
        if "next" in obj.attributes:
            b += obj.attributes["next"].to_bytes(4, "big")

        if "returnCode" in obj.attributes:
            b += obj.attributes["returnCode"].to_bytes(4, "big")

        return b

    def get_encoded_type(self, obj):
        type_id = self.get_type_id(obj.object_type)
        return type_id.to_bytes(1, "big")

    def encode(self, obj):
        message = bytes()
        message += self.get_encoded_type(obj)
        message += self.get_encoded_attributes(obj)
        message += bytes(obj.content, "UTF-8")
        return message


    def get_type_name(self, type_id):
        type_names = {
            1 : "text",
            2 : "loginAttempt",
            3 : "returnCode",
            4 : "command"
        }
        if type_id not in type_names:
            return "unknown"
        
        return type_names[type_id]

    def decode(self, s):
        obj = Object()
        obj.object_type = self.get_type_name(int.from_bytes(s[0:1], "big"))
        attributes_encoded = int.from_bytes(s[1:2], "big")
        
        ptr = 2
        if get_bit(attributes_encoded, 0):
            obj.attributes["id"] = int.from_bytes(s[ptr : (ptr + 4)], "big")
            ptr += 4

        if get_bit(attributes_encoded, 1):
            obj.attributes["author"] = int.from_bytes(s[ptr : (ptr + 4)], "big")
            ptr += 4

        if get_bit(attributes_encoded, 2):
            obj.attributes["timestamp"] = int.from_bytes(s[ptr : (ptr + 8)], "big")
            ptr += 8

        if get_bit(attributes_encoded, 3):
            obj.attributes["thread"] = int.from_bytes(s[ptr : (ptr + 4)], "big")
            ptr += 4

        if get_bit(attributes_encoded, 4):
            obj.attributes["reply"] = int.from_bytes(s[ptr : (ptr + 4)], "big")
            ptr += 4

        if get_bit(attributes_encoded, 5):
            obj.attributes["prev"] = int.from_bytes(s[ptr : (ptr + 4)], "big")
            ptr += 4

        if get_bit(attributes_encoded, 6):
            obj.attributes["next"] = int.from_bytes(s[ptr : (ptr + 4)], "big")
            ptr += 4

        if get_bit(attributes_encoded, 7):
            obj.attributes["returnCode"] = int.from_bytes(s[ptr : (ptr + 4)], "big")
            ptr += 4
        
        obj.content = s[ptr: ].decode("UTF-8")
        return obj